### PR TITLE
Fix uniformtext and enable coloraxis for sunburst and treemap as well as pathbar.textfont 

### LINKED ISSUE #4443: coloraxes with values for sunburst and treemap
Shared `coloraxes` should be implemented for `sunburst` and `treemap`.

### LINKED ISSUE #4451: pathbar.textfont has no effect in treemap
[Codepen](https://codepen.io/MojtabaSamimi/pen/OJPjqpj?editors=0010)

### LINKED ISSUE #4456: Problem fitting text inside bar-like traces when textangle is set
[Demo](https://codepen.io/MojtabaSamimi/pen/MWYEOLx?editors=0010)

### PR DESCRIPTION
**Pull request by archmoj**
Resolves #4443 by adding `coloraxis` to `sunburst` and `treemap` 
and fixes `uniformtext` scale issues with react (follow up of #4420):
- clear previous uniformtext scale before starting each plot type.
- fix uniformtext after switching levels when there is no transition for `sunburst` and `treemap`
- fix tests and add new tests for `uniformtext` as well as new `coloraxis` feature.

Also:
- fixes #4451 `pathbar.textfont` bug in  https://github.com/plotly/plotly.js/pull/4444/commits/6a4b451a9693898217d3fcaf6fb670b15a43734e
- fixes #4456 `textangle` bug in https://github.com/plotly/plotly.js/pull/4444/commits/2a7721c1d425a64b8a50f91734d5de1aea72c202

[demo treemap](https://codepen.io/MojtabaSamimi/pen/ExaXGQw)
[demo sunburst](https://codepen.io/MojtabaSamimi/pen/ExaXGbv)
[demo bar](https://codepen.io/MojtabaSamimi/pen/PowjRBa)

@plotly/plotly_js

### CODE REVIEW

File: src/traces/bar/plot.js [hunk 1]
```diff
@@ -91,6 +91,9 @@ function plot(gd, plotinfo, cdModule, traceLayer, opts, makeOnCompleteCallback)
             gap: fullLayout.bargap,
             groupgap: fullLayout.bargroupgap
         };
+
+        // don't clear bar when this is called from waterfall or funnel
+        clearMinTextSize('bar', fullLayout);
```
Comment (etpinard) [inline src/traces/bar/plot.js]: Wait. Here you're clearing the min-text-size stash during the plot step. So what's happens when a style-only edit (e.g. `Plotly.restyle(gd, 'marker.color', 'red')`) gets called?

To me, this logic should probably be somewhere in the calc step.

File: src/traces/bar/plot.js [hunk 2]
```diff
@@ -58,8 +58,8 @@ function getXY(di, xa, ya, isHorizontal) {
     return isHorizontal ? [s, p] : [p, s];
 }
 
-function transition(selection, opts, makeOnCompleteCallback) {
-    if(hasTransition(opts)) {
+function transition(selection, fullLayout, opts, makeOnCompleteCallback) {
+    if(!fullLayout.uniformtext.mode && hasTransition(opts)) {
```
Comment (etpinard) [inline src/traces/bar/plot.js]: Why? I thought you said this transitions looked fine

File: test/jasmine/tests/sunburst_test.js [hunk 3]
```diff
@@ -2062,31 +2062,103 @@ describe('sunburst uniformtext', function() {
...
+        })
+        .then(assertTextSizes('clear uniformtext', {
+            fontsizes: [12, 12, 12, 12, 12, 12, 12, 12, 12, 12],
+            scales: [1, 1, 1, 1, 1, 1, 1, 1, 1, 0.58],
+        }))
+        .catch(failTest)
+        .then(done);
+    });
+
+    it('should uniform text scales after transition', function(done) {
```
Comment (etpinard) [inline test/jasmine/tests/sunburst_test.js]: I'm getting failures on my laptop:

```
Chrome 78.0.3904 (Linux 0.0.0) sunburst uniformtext should uniform text scales after transition FAILED
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 0', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 1', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 7', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 9', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0, 1, 'scale for element 10', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 11', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 0', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 1', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 7', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 9', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0, 1, 'scale for element 10', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
	Error: Expected 0.86 to be close to 0.9725498243340049, 1, 'scale for element 11', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/sunburst_test.js:2012:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:226894:40
```

File: test/jasmine/tests/treemap_test.js [hunk 4]
```diff
@@ -1796,6 +1796,78 @@ describe('treemap uniformtext', function() {
...
+        })
+        .then(assertTextSizes('clear uniformtext', {
+            fontsizes: [12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12],
+            scales: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0.84],
+        }))
+        .catch(failTest)
+        .then(done);
+    });
+
+    it('should uniform text scales after transition', function(done) {
```
Comment (etpinard) [inline test/jasmine/tests/treemap_test.js]: Similarly here:

```
Chrome 78.0.3904 (Linux 0.0.0) treemap uniformtext should uniform text scales after transition FAILED
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 0', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 1', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 2', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 4', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 5', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0, 1, 'scale for element 7', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 9', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 10', 'before click'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.86 to be close to 0.9883899233296827, 1, 'scale for element 0', 'after click child'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.86 to be close to 0.9883899233296827, 1, 'scale for element 1', 'after click child'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.86 to be close to 0.9883899233296827, 1, 'scale for element 2', 'after click child'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.86 to be close to 0.9883899233296827, 1, 'scale for element 3', 'after click child'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.86 to be close to 0.9883899233296827, 1, 'scale for element 4', 'after click child'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.86 to be close to 0.9883899233296827, 1, 'scale for element 5', 'after click child'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.86 to be close to 0.7058823529411765, 1, 'scale for element 6', 'after click child'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 0', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 1', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 2', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 4', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 5', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0, 1, 'scale for element 7', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 9', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
	Error: Expected 0.84 to be close to 0.9424364123159304, 1, 'scale for element 10', 'after click parent'.
	    at <Jasmine>
	    at /tmp/test/jasmine/tests/treemap_test.js:1728:1 <- /tmp/51a75f08cee2dc3a8cecd7565be9a45d.browserify.js:228778:40
```

File: src/traces/sunburst/defaults.js [hunk 5]
```diff
@@ -44,7 +44,7 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
     if(lineWidth) coerce('marker.line.color', layout.paper_bgcolor);
 
     coerce('marker.colors');
-    var withColorscale = traceOut._hasColorscale = hasColorscale(traceIn, 'marker', 'colors');
+    var withColorscale = traceOut._hasColorscale = (traceIn.marker || {}).coloraxis || hasColorscale(traceIn, 'marker', 'colors');
```
Comment (etpinard) [inline src/traces/sunburst/defaults.js]: This here and in `treemap/defaults.js` would have been nicer as a separate commit.

----

Your new logic is difference than for other colorscale-compatible traces. See for example 

https://github.com/plotly/plotly.js/blob/e778644e1182e13126db6b801af6dd0a386372ba/src/traces/bar/style_defaults.js#L18-L22

- `hasColorscale(traceIn, 'marker', 'colors')` should be enough. For example, we should not coerce the colorscale attributes if `marker.coloraxis` is set, but `marker.colors` isn't a numerical array.

File: src/traces/bar/plot.js (hunk 2 above)
Comment (archmoj) [reply src/traces/bar/plot.js]: They did not apply a uniform text size during and after smooth transition.

File: src/traces/sunburst/defaults.js (hunk 5 above)
Comment (etpinard) [reply src/traces/sunburst/defaults.js]: cross-referencing https://github.com/plotly/plotly.js/issues/4443#issuecomment-569342008 - this new logic is :ok_hand: for sunburst and treemap traces that support colorscales w/o numerical `marker.colors`

---

Now @archmoj - as this is a special case, could you please add a comment about the

```
 var withColorscale = traceOut._hasColorscale = // ...
```

line saying something like:

```
// N.B. special logic to consider "values" colorscales
```

Comment (archmoj) [reply src/traces/sunburst/defaults.js]: Done in https://github.com/plotly/plotly.js/pull/4444/commits/ef6a95258c8cb83842dbabe7be13a85e1606cf75.

File: src/traces/bar/plot.js (hunk 1 above)
Comment (archmoj) [reply src/traces/bar/plot.js]: Restyle works. Considering interactions (e.g. zoom and selection) this should be cleared after calc and at the stat of plot.

File: test/jasmine/tests/treemap_test.js (hunk 4 above)
Comment (archmoj) [reply test/jasmine/tests/treemap_test.js]: It should be improved in https://github.com/plotly/plotly.js/pull/4444/commits/2cf7defe563cbdaac46d880e6e54aa6f21965888.

File: test/jasmine/tests/sunburst_test.js (hunk 3 above)
Comment (archmoj) [reply test/jasmine/tests/sunburst_test.js]: It should be improved in https://github.com/plotly/plotly.js/pull/4444/commits/2cf7defe563cbdaac46d880e6e54aa6f21965888.

File: src/traces/bar/plot.js (hunk 1 above)
Comment (etpinard) [reply src/traces/bar/plot.js]: > Considering interactions (e.g. zoom and selection) this should be cleared after calc and at the stat of plot.

Can you explain a bit more here? To me, sounds like we should not clear `minTextSize` during zoom interactions.

File: src/traces/sunburst/defaults.js (hunk 5 above)
Comment (etpinard) [reply src/traces/sunburst/defaults.js]: Nicely done. Thanks!!

File: test/jasmine/tests/treemap_test.js (hunk 4 above)
Comment (etpinard) [reply test/jasmine/tests/treemap_test.js]: Nicely done. All tests pass now :ok_hand:

File: test/jasmine/tests/sunburst_test.js (hunk 3 above)
Comment (etpinard) [reply test/jasmine/tests/sunburst_test.js]: Now passing :1st_place_medal:

File: src/traces/treemap/draw_ancestors.js [hunk 6]
```diff
@@ -141,7 +141,9 @@ module.exports = function drawAncestors(gd, cd, entry, slices, opts) {
             s.attr('data-notex', 1);
         });
 
-        var font = Lib.ensureUniformFontSize(gd, helpers.determineTextFont(trace, pt, fullLayout.font, trace.pathdir));
+        var font = Lib.ensureUniformFontSize(gd, helpers.determineTextFont(trace, pt, fullLayout.font, {
+            onPathbar: true
```
Comment (etpinard) [inline src/traces/treemap/draw_ancestors.js]: Can you explain why this commit made three existing mocks change (`treemap_level-depth`, `treemap_packings` and `treemap_textposition`) ?

Comment (archmoj) [reply src/traces/treemap/draw_ancestors.js]: Good question. 
And a bad mistake here. 
There is no such a thing as `pathdir`!
It should have been written `pathbar`.

File: src/traces/bar/plot.js (hunk 1 above)
Comment (etpinard) [reply src/traces/bar/plot.js]: Or let me rephrase my question: if you do move `clearMinTextSize` to the calc step, do any of the tests fail?

Comment (archmoj) [reply src/traces/bar/plot.js]: Yes. In that case the react tests would fail.

**Comment by archmoj** [comment]
After https://github.com/plotly/plotly.js/pull/4444/commits/a9c958942fbbd2bafbeeee386bf734ec3f25f9eb this PR is now ready to go.

File: src/traces/bar/plot.js (hunk 1 above)
Comment (etpinard) [reply src/traces/bar/plot.js]: Can you share the branch you used to test that?

File: test/jasmine/tests/bar_test.js [hunk 7]
```diff
@@ -2877,7 +2877,7 @@ describe('bar uniformtext', function() {
         })
         .then(assertTextSizes('using minsize: 9', {
             fontsizes: [12, 12, 12, 12, 12, 12, 12],
-            scales: [0, 0.48, 0.48, 0.48, 0.48, 0.48, 0.48],
+            scales: [0, 1, 1, 1, 1, 1, 1],
```
Comment (etpinard) [inline test/jasmine/tests/bar_test.js]: Why did this change?

Comment (archmoj) [reply test/jasmine/tests/bar_test.js]: Used to computed wrong scales as the minsize was not cleared.

Comment (etpinard) [reply test/jasmine/tests/bar_test.js]: Great. Thanks!

File: src/traces/bar/plot.js (hunk 1 above)
Comment (etpinard) [reply src/traces/bar/plot.js]: Summing up a private convo:

- I was under the impression that the `uniformtext` attributes were `editType: 'calc'`, so that's why I thought it would be best to clear the min-text-size during the calc step, but they're not. The `uniformtext` attributes are `editType: 'plot'`.
- @archmoj says that making the `uniformtext` attributes `editType: 'calc'` would have an impact on the way graph with set uniformtext would behave on zoom
- Since we'll need to refactor the trace text pipeline at some point (more info in https://github.com/plotly/plotly.js/pull/4420#issuecomment-565189036), let's keep the clear-min-text logic in the plot step for now.
- The most important of this PR are the newly added tests.

**Comment by archmoj** [comment]
Here is info related to new/modified baselines:
textangle bug fix (#4456): 
- (mod) funnel_axis_textangle
- (mod) funnel_axis_textangle_start-end
- (new) bar-like_textangle45 
- (new) bar-like_textangle60

treemap trace custom fonts bug fix (#4451)
- (new) treemap_fonts_nocolor
- (new) treemap_fonts_withcolor 
- (mod) treemap_packings
- (mod) treemap_textposition
- (mod) uniformtext_sunburst_treemap

treemap coloraxis with values as well as custom fonts and uniformtext (#4443):
- (new) uniformtext_treemap

uniform start/end anchor positions in respect to (#4247):
- (mod) niformtext_bar_edgecase2 - text positioned with an identical pad from the top of the bar
- (mod) uniformtext_bar_edgecase3 - text positioned with an identical pad from the top of the bar 
- (new) uniformtext_bar_edgecase4 - text positioned with an identical pad from the bottom of the bar
- (new) uniformtext_bar_edgecase5 - text positioned with an identical pad from the bottom of the bar - case of tilted text
- (new) uniformtext_bar_edgecase5 - text positioned with an identical pad from the bottom of the bar - case of vertical text on vertical bars
- (new) uniformtext_bar_edgecase6 - text positioned with an identical pad from the bottom of the bar - case of horizontal text on horizontal bars
- (new) uniformtext_bar_edgecase6 - text positioned with an identical pad from the bottom of the bar - case of tilted text on horizontal bars
- (new) uniformtext_bar_axis_textangle_inside - inside text positioned with an identical pad in different axis orientations
- (new) uniformtext_bar_axis_textangle_outside - inside text positioned with an identical pad in different axis orientations

modified mock: middle anchors is not used now that the text positions properly by default 
- (mod) uniformtext_bar-like_10_auto
- (mod) uniformtext_bar-like_8_horizontal
- (mod) uniformtext_bar-like_8_textangle
- (mod) uniformtext_bar-like_8_textangle45

**Comment by etpinard** [comment]
:dancer: :dancer:
